// hrr-memory-obs/src/conflict.js
/**
 * ConflictDetector — algebraic conflict detection using HRR symbol vectors.
 *
 * @typedef {{ ts: number, subject: string, relation: string, newObject: string, oldObject: string, similarity: number }} ConflictFlag
 */

import { similarity } from 'hrr-memory';

export class ConflictDetector {
  constructor(hrr, threshold = 0.3) {
    this._hrr = hrr;
    this._threshold = threshold;
    this._knownPairs = new Set(); // track subject+relation pairs we've seen
  }

  /**
   * Record that a (subject, relation) pair has been stored.
   * Call this after a successful store so future checks know to run the query.
   */
  track(subject, relation) {
    this._knownPairs.add(`${subject}\0${relation}`);
  }

  check(subject, relation, object, ts) {
    const s = subject.toLowerCase().trim();
    const r = relation.toLowerCase().trim();

    // Fast path: if we've never stored this pair, no conflict possible.
    // Skip the expensive HRR query entirely.
    if (!this._knownPairs.has(`${s}\0${r}`)) return null;

    const existing = this._hrr.query(subject, relation);
    if (!existing.confident) return null;

    const oldVec = this._hrr.symbols.get(existing.match);
    const newVec = this._hrr.symbols.get(object.toLowerCase().trim());
    const sim = similarity(oldVec, newVec);

    if (sim >= this._threshold) return null;

    return {
      ts: ts ?? Date.now(),
      subject: subject.toLowerCase().trim(),
      relation: relation.toLowerCase().trim(),
      newObject: object.toLowerCase().trim(),
      oldObject: existing.match,
      similarity: Math.round(sim * 1000) / 1000,
    };
  }
}
