// hrr-memory-obs/src/observation.js
/**
 * ObservationMemory — composition wrapper around HRRMemory.
 *
 * Adds timeline recording, conflict detection, and observation consolidation
 * on top of HRRMemory's triple store.
 *
 * @typedef {{ id: string, subject: string, observation: string, evidence: Array<{ ts: number, triple: [string, string, string] }>, confidence: 'high' | 'medium' | 'low', supersedes: string[], createdAt: number }} Observation
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { HRRMemory } from 'hrr-memory';
import { Timeline } from './timeline.js';
import { ConflictDetector } from './conflict.js';
import { defaultPrompt } from './prompt.js';

export class ObservationParseError extends Error {
  constructor(message, raw) {
    super(message);
    this.name = 'ObservationParseError';
    this.raw = raw;
  }
}

export class ObservationMemory {
  constructor(hrr, options = {}) {
    this._hrr = hrr;
    this._executor = options.executor || null;
    this._autoConsolidateAfter = options.autoConsolidateAfter ?? null;
    this._promptFn = options.promptFn || defaultPrompt;
    this._timeline = new Timeline();
    this._conflict = new ConflictDetector(hrr, options.conflictThreshold ?? 0.3);
    this._flags = [];
    this._observations = [];
    this._obsCounter = 0;
    this._consolidating = false;
    this._meta = {
      createdAt: Date.now(),
      lastConsolidation: null,
      totalStores: 0,
      totalForgets: 0,
      totalConsolidations: 0,
    };
  }

  // ── Delegated methods ──────────────────────────────

  query(subject, relation) { return this._hrr.query(subject, relation); }
  querySubject(subject) { return this._hrr.querySubject(subject); }
  search(relation, object) { return this._hrr.search(relation, object); }
  ask(question) { return this._hrr.ask(question); }
  stats() { return this._hrr.stats(); }

  // ── Store (async) ──────────────────────────────────

  async store(subject, relation, object) {
    const ts = Date.now();
    const s = subject.toLowerCase().trim();
    const r = relation.toLowerCase().trim();
    const o = object.toLowerCase().trim();

    // Check for conflict BEFORE storing (so query sees old state)
    const flag = this._conflict.check(s, r, o, ts);

    // Delegate to HRR
    const stored = this._hrr.store(subject, relation, object);

    // Only record timeline entry if actually stored (not a duplicate)
    if (stored) {
      const entry = { ts, subject: s, relation: r, object: o, op: 'store' };
      if (flag) {
        entry.conflict = { oldObject: flag.oldObject, similarity: flag.similarity };
        this._flags.push(flag);
      }
      this._timeline.append(entry);
      this._meta.totalStores++;

      // Track this pair so future stores trigger conflict checks
      this._conflict.track(s, r);

      // Auto-consolidation
      if (this._autoConsolidateAfter && this._flags.length >= this._autoConsolidateAfter) {
        await this.consolidate();
      }
    }

    return stored;
  }

  // ── Forget (async) ─────────────────────────────────

  async forget(subject, relation, object) {
    const ts = Date.now();
    const s = subject.toLowerCase().trim();
    const r = relation.toLowerCase().trim();
    const o = object.toLowerCase().trim();

    const forgotten = this._hrr.forget(subject, relation, object);

    if (forgotten) {
      this._timeline.append({ ts, subject: s, relation: r, object: o, op: 'forget' });
      this._meta.totalForgets++;
    }

    return forgotten;
  }

  // ── Timeline ───────────────────────────────────────

  history(subject, relation) { return this._timeline.history(subject, relation); }
  at(ts) { return this._timeline.at(ts); }

  // ── Flags ──────────────────────────────────────────

  flags() { return [...this._flags]; }

  // ── Observations ───────────────────────────────────

  observations(subject) {
    let obs = [...this._observations];
    if (subject) {
      obs = obs.filter(o => o.subject === subject.toLowerCase().trim());
    }
    return obs.sort((a, b) => b.createdAt - a.createdAt);
  }

  /**
   * Add an observation directly (without calling an LLM).
   * Use this when an external agent has already synthesized the observation.
   * @param {{ subject: string, observation: string, evidence: Array<{ ts: number, triple: [string, string, string] }>, confidence: 'high' | 'medium' | 'low', supersedes?: string[] }} obs
   * @returns {Observation}
   */
  addObservation({ subject, observation, evidence, confidence, supersedes = [] }) {
    const obs = {
      id: `obs_${String(++this._obsCounter).padStart(3, '0')}`,
      subject: subject.toLowerCase().trim(),
      observation,
      evidence,
      confidence,
      supersedes,
      createdAt: Date.now(),
    };
    this._observations.push(obs);
    return obs;
  }

  /**
   * Clear conflict flags for a specific subject.
   * @param {string} subject
   */
  clearFlags(subject) {
    const s = subject.toLowerCase().trim();
    this._flags = this._flags.filter(f => f.subject !== s);
  }

  async consolidate() {
    if (!this._executor) {
      throw new Error('No executor configured. Pass executor in ObservationMemory options.');
    }
    if (this._flags.length === 0) return [];
    if (this._consolidating) return [];

    this._consolidating = true;
    try {
      // Gather timeline entries for flagged subjects
      const subjects = new Set(this._flags.map(f => f.subject));
      const entries = [];
      for (const s of subjects) {
        entries.push(...this._timeline.history(s));
      }

      // Gather existing observations for flagged subjects
      const existingObservations = this._observations
        .filter(o => subjects.has(o.subject))
        .map(o => ({ id: o.id, subject: o.subject, observation: o.observation, confidence: o.confidence }));

      // Build prompt
      const input = { entries, flags: [...this._flags], existingObservations };
      const prompt = this._promptFn(input);

      // Call LLM
      const raw = await this._executor(prompt);

      // Parse response
      let parsed;
      try {
        parsed = JSON.parse(raw);
      } catch {
        throw new ObservationParseError('Executor returned invalid JSON', raw);
      }

      if (!parsed.observations || !Array.isArray(parsed.observations)) {
        throw new ObservationParseError('Response missing "observations" array', raw);
      }

      // Validate and store observations
      const allTimestamps = new Set(this._timeline.all().map(e => e.ts));
      const allObsIds = new Set(this._observations.map(o => o.id));
      const newObs = [];

      for (const raw of parsed.observations) {
        // Require fields
        if (!raw.subject || !raw.observation || !raw.evidence || !raw.confidence) continue;

        // Filter evidence to valid timestamps
        const validEvidence = (Array.isArray(raw.evidence) ? raw.evidence : [])
          .filter(e => e.ts && allTimestamps.has(e.ts) && Array.isArray(e.triple));

        if (validEvidence.length === 0) continue;

        // Filter supersedes to valid IDs
        const validSupersedes = (Array.isArray(raw.supersedes) ? raw.supersedes : [])
          .filter(id => allObsIds.has(id));

        const obs = {
          id: `obs_${String(++this._obsCounter).padStart(3, '0')}`,
          subject: raw.subject,
          observation: raw.observation,
          evidence: validEvidence,
          confidence: raw.confidence,
          supersedes: validSupersedes,
          createdAt: Date.now(),
        };

        this._observations.push(obs);
        allObsIds.add(obs.id);
        newObs.push(obs);
      }

      // Clear flags
      this._flags = [];
      this._meta.lastConsolidation = Date.now();
      this._meta.totalConsolidations++;

      return newObs;
    } finally {
      this._consolidating = false;
    }
  }

  // ── Persistence ────────────────────────────────────

  toJSON() {
    return {
      version: 1,
      meta: { ...this._meta },
      timeline: this._timeline.toJSON(),
      observations: this._observations.map(o => ({ ...o })),
      flags: this._flags.map(f => ({ ...f })),
      obsCounter: this._obsCounter,
    };
  }

  static fromJSON(hrr, data, options = {}) {
    const mem = new ObservationMemory(hrr, options);
    if (data.meta) mem._meta = { ...data.meta };
    if (data.timeline) mem._timeline = Timeline.fromJSON(data.timeline);
    if (data.observations) mem._observations = data.observations.map(o => ({ ...o }));
    if (data.flags) mem._flags = data.flags.map(f => ({ ...f }));
    if (typeof data.obsCounter === 'number') mem._obsCounter = data.obsCounter;
    // Rebuild known pairs from timeline so conflict detection works after load
    if (data.timeline) {
      for (const e of data.timeline) {
        if (e.op === 'store') mem._conflict.track(e.subject, e.relation);
      }
    }
    return mem;
  }

  save(hrrPath, obsPath) {
    this._hrr.save(hrrPath);
    writeFileSync(obsPath, JSON.stringify(this.toJSON(), null, 2));
  }

  static load(hrrPath, obsPath, options = {}) {
    const hrr = HRRMemory.load(hrrPath);
    if (!existsSync(obsPath)) return new ObservationMemory(hrr, options);
    try {
      const data = JSON.parse(readFileSync(obsPath, 'utf8'));
      return ObservationMemory.fromJSON(hrr, data, options);
    } catch {
      return new ObservationMemory(hrr, options);
    }
  }
}
