// hrr-memory-obs/src/timeline.js
/**
 * Timeline — append-only event log for HRR memory operations.
 *
 * @typedef {{ ts: number, subject: string, relation: string, object: string, op: 'store' | 'forget', conflict?: { oldObject: string, similarity: number } }} TimelineEntry
 */

export class Timeline {
  constructor() {
    /** @type {TimelineEntry[]} */
    this._entries = [];
  }

  append(entry) {
    this._entries.push({ ...entry });
  }

  history(subject, relation) {
    let results = this._entries.filter(e => e.subject === subject);
    if (relation !== undefined) {
      results = results.filter(e => e.relation === relation);
    }
    return results.sort((a, b) => a.ts - b.ts);
  }

  all() {
    return [...this._entries].sort((a, b) => a.ts - b.ts);
  }

  get length() {
    return this._entries.length;
  }

  at(ts) {
    return new SymbolicProxy(this._entries, ts);
  }

  toJSON() {
    return this._entries.map(e => ({ ...e }));
  }

  static fromJSON(data) {
    const tl = new Timeline();
    for (const entry of data) {
      tl._entries.push({ ...entry });
    }
    return tl;
  }
}

export class SymbolicProxy {
  constructor(entries, ts) {
    this._entries = entries;
    this._ts = ts;
  }

  facts(subject, relation) {
    const relevant = this._entries
      .filter(e => e.ts <= this._ts && e.subject === subject
        && (relation === undefined || e.relation === relation))
      .sort((a, b) => a.ts - b.ts);

    // Use Map<relation\0object, object> to handle colons in values
    const live = new Map();
    for (const e of relevant) {
      const key = `${e.relation}\0${e.object}`;
      if (e.op === 'store') live.set(key, e.object);
      else if (e.op === 'forget') live.delete(key);
    }
    return [...live.values()];
  }
}
